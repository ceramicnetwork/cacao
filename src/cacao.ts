import { SiweMessage } from 'siwe'

export type Header = {
  t: string
}

export type Payload = {
  domain: string
  iss: string
  chainId: number
  aud: string
  version: string
  nonce: string
  iat: number
  nbf?: number
  exp?: number
  statement?: string
  requestId?: string
  resources?: string[]
}

export type Signature = {
  s: string
}

export type Cacao = {
  h: Header
  p: Payload
  s: Signature
}

export interface CacaoBlock {
  cacao: Cacao
  // cid: CID
  bytes: Uint8Array
}

class CACAO {
  cacao: Cacao
  constructor(siwe: string | SiweMessage) {
    const siweMessage = typeof siwe === 'string' ? new SiweMessage(siwe) : siwe
    this.cacao = this.fromSiweMessage(siweMessage)
  }

  // TODO: Implement this API
  // toCacaoBlock(): CacaoBlock {}

  private fromSiweMessage(siweMessage: SiweMessage): Cacao {
    return {
      h: {
        t: 'eip4361-eip191',
      },
      p: {
        domain: siweMessage.domain,
        exp: Date.parse(siweMessage.expirationTime),
        iat: Date.parse(siweMessage.issuedAt),
        iss: siweMessage.address,
        chainId: Number(siweMessage.chainId),
        aud: siweMessage.uri,
        version: siweMessage.version,
        nonce: siweMessage.nonce,
        nbf: Date.parse(siweMessage.notBefore),
        statement: siweMessage.statement,
        requestId: siweMessage.requestId,
        resources: siweMessage.resources,
      },
      s: {
        s: siweMessage.signature,
      },
    }
  }
}

export default CACAO
