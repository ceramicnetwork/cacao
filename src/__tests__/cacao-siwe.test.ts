import { Wallet } from '@ethersproject/wallet'
import { Cacao, CacaoBlock } from '../cacao.js'
import { SiweMessage } from '../siwx/siwe.js'

const ISSUED_AT = new Date('2021-10-14T07:18:41Z')

const WALLET_MNEMONIC =
  'despair voyage estate pizza main slice acquire mesh polar short desk lyrics'
const ETHEREUM_WALLET = Wallet.fromMnemonic(WALLET_MNEMONIC)
const ETHEREUM_ADDRESS = ETHEREUM_WALLET.address
const SIWE_MESSAGE_PARAMS = {
  domain: 'service.org',
  address: ETHEREUM_ADDRESS,
  statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
  uri: 'did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8',
  version: '1',
  nonce: '32891757',
  issuedAt: ISSUED_AT.toISOString(),
  chainId: '1',
  resources: [
    'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
    'https://example.com/my-web2-claim.json',
  ],
}

test('Can create and verify Cacao Block for Ethereum', async () => {
  const msg = new SiweMessage(SIWE_MESSAGE_PARAMS)

  msg.signature = await ETHEREUM_WALLET.signMessage(msg.signMessage())

  const cacao = Cacao.fromSiweMessage(msg)
  const block = await CacaoBlock.fromCacao(cacao)
  expect(block).toMatchSnapshot()

  expect(() => Cacao.verify(cacao)).not.toThrow()
})

test('Converts between Cacao and SiweMessage', () => {
  const msg = new SiweMessage(SIWE_MESSAGE_PARAMS)

  const cacao = Cacao.fromSiweMessage(msg)
  const siwe = SiweMessage.fromCacao(cacao)
  expect(siwe).toEqual(msg)
})

test('ok after exp if within phase out period', async () => {
  const msg = new SiweMessage({
    ...SIWE_MESSAGE_PARAMS,
    expirationTime: new Date(ISSUED_AT.valueOf() + 5 * 1000).toISOString(),
  })

  msg.signature = await ETHEREUM_WALLET.signMessage(msg.toMessage())

  const cacao = Cacao.fromSiweMessage(msg)
  const expiredTime = new Date(ISSUED_AT.valueOf() + 10 * 1000)
  expect(() =>
    Cacao.verify(cacao, { atTime: expiredTime, revocationPhaseOutSecs: 20, clockSkewSecs: 0 })
  ).not.toThrow()
})

test('fail after exp if after phase out period', async () => {
  const msg = new SiweMessage({
    ...SIWE_MESSAGE_PARAMS,
    expirationTime: new Date(ISSUED_AT.valueOf() + 5 * 1000).toISOString(),
  })

  msg.signature = await ETHEREUM_WALLET.signMessage(msg.signMessage())

  const cacao = Cacao.fromSiweMessage(msg)
  const expiredTime = new Date(ISSUED_AT.valueOf() + 10 * 1000)
  expect(() =>
    Cacao.verify(cacao, { atTime: expiredTime, revocationPhaseOutSecs: 1, clockSkewSecs: 0 })
  ).toThrow(`CACAO has expired`)
})

test('ok before IAT if within default clockskew', async () => {
  const msg = new SiweMessage(SIWE_MESSAGE_PARAMS)

  msg.signature = await ETHEREUM_WALLET.signMessage(msg.toMessage())

  const cacao = Cacao.fromSiweMessage(msg)
  const OneMinbeforeIAT = new Date(ISSUED_AT.valueOf() - 60 * 1000)
  expect(() => Cacao.verify(cacao, { atTime: OneMinbeforeIAT })).not.toThrow()
})

test('ok after exp if disableTimecheck option', async () => {
  const msg = new SiweMessage({
    ...SIWE_MESSAGE_PARAMS,
    expirationTime: new Date(ISSUED_AT.valueOf() + 5 * 1000).toISOString(),
  })

  msg.signature = await ETHEREUM_WALLET.signMessage(msg.toMessage())

  const cacao = Cacao.fromSiweMessage(msg)
  expect(() =>
    Cacao.verify(cacao, {
      disableExpirationCheck: true,
      revocationPhaseOutSecs: 20,
      clockSkewSecs: 0,
    })
  ).not.toThrow()
})
