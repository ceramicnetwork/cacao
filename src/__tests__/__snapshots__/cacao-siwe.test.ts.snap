// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cacao SIWE Can create and verify Cacao Block for Ethereum 1`] = `
Block {
  "asBlock": [Circular],
  "bytes": Uint8Array [
    163,
    97,
    104,
    161,
    97,
    116,
    103,
    99,
    97,
    105,
    112,
    49,
    50,
    50,
    97,
    112,
    169,
    99,
    97,
    117,
    100,
    120,
    56,
    100,
    105,
    100,
    58,
    107,
    101,
    121,
    58,
    122,
    54,
    77,
    107,
    114,
    66,
    100,
    78,
    100,
    119,
    85,
    80,
    110,
    88,
    68,
    86,
    68,
    49,
    68,
    67,
    120,
    101,
    100,
    122,
    86,
    86,
    66,
    112,
    97,
    71,
    105,
    56,
    97,
    83,
    109,
    111,
    88,
    70,
    65,
    101,
    75,
    78,
    103,
    116,
    65,
    101,
    114,
    56,
    99,
    101,
    120,
    112,
    120,
    24,
    50,
    48,
    50,
    49,
    45,
    49,
    48,
    45,
    51,
    48,
    84,
    49,
    54,
    58,
    50,
    53,
    58,
    50,
    52,
    46,
    48,
    48,
    48,
    90,
    99,
    105,
    97,
    116,
    120,
    24,
    50,
    48,
    50,
    49,
    45,
    48,
    57,
    45,
    51,
    48,
    84,
    49,
    54,
    58,
    50,
    53,
    58,
    50,
    52,
    46,
    48,
    48,
    48,
    90,
    99,
    105,
    115,
    115,
    120,
    59,
    100,
    105,
    100,
    58,
    112,
    107,
    104,
    58,
    101,
    105,
    112,
    49,
    53,
    53,
    58,
    49,
    58,
    48,
    120,
    66,
    100,
    57,
    68,
    57,
    99,
    55,
    68,
    67,
    51,
    56,
    57,
    55,
    49,
    53,
    97,
    56,
    57,
    102,
    67,
    56,
    49,
    52,
    57,
    69,
    52,
    97,
    53,
    66,
    101,
    57,
    49,
    51,
    51,
    54,
    66,
    50,
    55,
    57,
    54,
    101,
    110,
    111,
    110,
    99,
    101,
    104,
    51,
    50,
    56,
    57,
    49,
    55,
    53,
    55,
    102,
    100,
    111,
    109,
    97,
    105,
    110,
    107,
    115,
    101,
    114,
    118,
    105,
    99,
    101,
    46,
    111,
    114,
    103,
    103,
    118,
    101,
    114,
    115,
    105,
    111,
    110,
    97,
    49,
    105,
    114,
    101,
    115,
    111,
    117,
    114,
    99,
    101,
    115,
    130,
    120,
    53,
    105,
    112,
    102,
    115,
    58,
    47,
    47,
    81,
    109,
    101,
    55,
    115,
    115,
    51,
    65,
    82,
    86,
    103,
    120,
    118,
    54,
    114,
    88,
    113,
    86,
    80,
    105,
    105,
    107,
    77,
    74,
    56,
    117,
    50,
    78,
    76,
    103,
    109,
    103,
    115,
    122,
    103,
    49,
    51,
    112,
    89,
    114,
    68,
    75,
    69,
    111,
    105,
    117,
    120,
    38,
    104,
    116,
    116,
    112,
    115,
    58,
    47,
    47,
    101,
    120,
    97,
    109,
    112,
    108,
    101,
    46,
    99,
    111,
    109,
    47,
    109,
    121,
    45,
    119,
    101,
    98,
    50,
    45,
    99,
    108,
    97,
    105,
    109,
    46,
    106,
    115,
    111,
    110,
    105,
    115,
    116,
    97,
    116,
    101,
    109,
    101,
    110,
    116,
    120,
    65,
    73,
    32,
    97,
    99,
    99,
    101,
    112,
    116,
    32,
    116,
    104,
    101,
    32,
    83,
    101,
    114,
    118,
    105,
    99,
    101,
    79,
    114,
    103,
    32,
    84,
    101,
    114,
    109,
    115,
    32,
    111,
    102,
    32,
    83,
    101,
    114,
    118,
    105,
    99,
    101,
    58,
    32,
    104,
    116,
    116,
    112,
    115,
    58,
    47,
    47,
    115,
    101,
    114,
    118,
    105,
    99,
    101,
    46,
    111,
    114,
    103,
    47,
    116,
    111,
    115,
    97,
    115,
    162,
    97,
    115,
    120,
    132,
    48,
    120,
    99,
    51,
    49,
    48,
    57,
    98,
    102,
    48,
    98,
    101,
    100,
    101,
    97,
    57,
    51,
    99,
    100,
    55,
    57,
    57,
    97,
    98,
    57,
    49,
    100,
    98,
    53,
    56,
    100,
    52,
    100,
    54,
    50,
    98,
    51,
    99,
    100,
    56,
    53,
    97,
    99,
    57,
    97,
    48,
    56,
    97,
    51,
    97,
    100,
    100,
    56,
    57,
    52,
    52,
    52,
    51,
    98,
    55,
    100,
    52,
    48,
    55,
    101,
    49,
    49,
    51,
    56,
    98,
    52,
    50,
    49,
    100,
    55,
    98,
    100,
    50,
    49,
    52,
    55,
    100,
    102,
    99,
    54,
    49,
    57,
    99,
    54,
    48,
    102,
    54,
    100,
    48,
    100,
    48,
    102,
    101,
    54,
    54,
    51,
    101,
    98,
    101,
    53,
    57,
    53,
    51,
    57,
    102,
    57,
    99,
    50,
    101,
    48,
    100,
    98,
    57,
    51,
    99,
    49,
    56,
    54,
    49,
    51,
    56,
    102,
    48,
    97,
    98,
    49,
    99,
    97,
    116,
    102,
    101,
    105,
    112,
    49,
    57,
    49,
  ],
  "cid": Object {
    "code": 113,
    "hash": Uint8Array [
      18,
      32,
      81,
      49,
      11,
      96,
      114,
      232,
      123,
      166,
      247,
      231,
      246,
      53,
      176,
      107,
      5,
      233,
      68,
      245,
      50,
      61,
      162,
      56,
      245,
      54,
      101,
      143,
      102,
      183,
      28,
      4,
      211,
      193,
    ],
    "version": 1,
  },
  "value": Object {
    "h": Object {
      "t": "caip122",
    },
    "p": Object {
      "aud": "did:key:z6MkrBdNdwUPnXDVD1DCxedzVVBpaGi8aSmoXFAeKNgtAer8",
      "domain": "service.org",
      "exp": "2021-10-30T16:25:24.000Z",
      "iat": "2021-09-30T16:25:24.000Z",
      "iss": "did:pkh:eip155:1:0xBd9D9c7DC389715a89fC8149E4a5Be91336B2796",
      "nonce": "32891757",
      "resources": Array [
        "ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu",
        "https://example.com/my-web2-claim.json",
      ],
      "statement": "I accept the ServiceOrg Terms of Service: https://service.org/tos",
      "version": "1",
    },
    "s": Object {
      "s": "0xc3109bf0bedea93cd799ab91db58d4d62b3cd85ac9a08a3add894443b7d407e1138b421d7bd2147dfc619c60f6d0d0fe663ebe59539f9c2e0db93c186138f0ab1c",
      "t": "eip191",
    },
  },
}
`;
